nnvm/python/nnvm/frontend/tensorflow.py

```diff
@@ -1279,6 +1307,17 @@ def from_tensorflow(self, graph, layout="NHWC", shape=None, outputs=None):
                 self._nodes[node.name] = op
                 final_op = op
 
+                # Infer shapes even without specifying "add_shapes=True"
+                if output_shapes == [None]:
+                    g = _graph.create(final_op)
+                    self._output_shapes[node.name] = \
+                        list(graph_util.infer_shape(g, **self._input_shapes))[-1]
+                else:
+                    self._output_shapes[node.name] = output_shapes
+
+                if self._output_shapes[node.name] and shape and node.name in shape:
+                    assert self._input_shapes[node.name] == list(shape[node.name])
+
             # Infer shapes if passed explicitely
             node_output = self._nodes[node.name]
             if shape:
```

Comment:
I think this if block is redundant now, can be removed.

Reply:
after thinking about it for a while, I'd like to keep this block, but restrict the if condition.
- the inferred shape above is [[]] or [[],[],[],...] 
- -1 is in the inferred shape

if any of the two cases happen for some reason, then let's try to infer again using user's shape. This helped pass test case test_forward_unpack as well.

Reply:
mmm, with all enhancements above this should never occur.
Ok, let this fallback be there for now.